### PR TITLE
Refactor data services to use composition

### LINKED ISSUE #742: Cache redesign part 4: Split up the dataservices
## References
This ticket corresponds to issue 8 from #718. 

That ticket was too big to tackle at once so we split it up in to smaller, more manageable chunks of work.

## Description
We encourage every service that corresponds with a rest endpoint to inherit from it. But most endpoints only support a fraction of the methods dataservice provides. So we either inherit, and a lot of methods don't work, or use delegation, but don't have interfaces to enforce which methods the delegating service has. 

This makes it possible for someone to add an `@dataservice` annotation on a service that isn't one, and we'd only know about it at runtime. We should at least create a separate interface for each atomic bit of functionality in dataservice, and perhaps also split up dataservice in to multiple smaller classes, and use composition rather than inheritance for the implementing services.

## Proposed Solution

- We'll split dataservice methods up in to chunks of functionality and create interfaces for them. 
- e.g. CreateDataService, ReadDataService, PutDataService, PatchDataService, DeleteDataService, SearchDataService, … 
- The dataservice implementations don't inherit from dataservice but implement the interfaces for the methods they need
- So DSODataservice would only implement ReadDataService, while CommunityDataService would implement them all.
- What is currently the @DataService annotation would become @ReadDataService and enforce services using it to implement the ReadDataService interface.

This will take an estimated **24 - 32h**

### PR DESCRIPTION
**Pull request by ybnd**
## References
* Fixes #742

## Description
This PR refactors all data services into a composition pattern

* Data services should extend `BaseDataService` or `IdentifiableDataService`
  * `BaseDataService` contains only the most basic methods shared by all
  * `IdentifiableDataService` also contains methods that address resources by their ID. We chose to use inheritance instead of composition because of the number of services that depend on IDs.
* All other `DataService` functionality was split up into "feature" classes, that services can now compose. 
  * The following feature classes are available:
    * `CreateData`
    * `FindAllData`
    * `SearchData`
    * `PatchData`
    * `PutData`
    * `DeleteData`
  * To support one of these features, a concrete data service class must
    * implement the feature's interface (e.g. `SearchData`)
    * wrap an instance of that feature's implementation (e.g. `SearchDataImpl`)
    * forward the interface's methods to that instance
  * Some logic that used to be modified on a case-by-case basis by overrides had to be reworked
    * "ID endpoints" can be adjusted by passing a `constructIdEndpoint` function. In this way they can be synchronised between all instances a data service uses.
    * "search endpoints" are handled in the same way so we can just reuse `SearchDataImpl` everywhere
    
* A special case was made for`HrefDataService`: it now uses a "full" `BaseDataService` under the hood, but only exposes `findByHref` and `findListByHref` while changing their signature to support
  ```
  itemRD$: Observable<RemoteData<Item>> = hrefOnlyDataService.findByHref<Item>(...);
  ```
* The `@dataService` decorator is now typed more strictly
  * For the purposes of `LinkService`, it is sufficient for data services to implement  `findByHref` and `findListByHref` 
    `BaseDataService` and `HrefOnlyDataService` implement a common interface that only contains these two methods: `HALDataService`
  * `@dataService` can only decorate classes that implement `HALDataService`
  * This also means that `DATA_SERVICE_FACTORY` can now guarantee that all data services can 

## Instructions for Reviewers
* Go through the app and test as many CRUD operations as is reasonable

  * Browse / search / DSO pages
  * Submission
  * Workflow
  * EPerson / Group administration
  * Registries
  * ...

  Everything should be functionally the same as it was before this PR, no unexpected errors should come up.

* `@dataService` decorators on classes that _don't_ implement `HALDataService` should now be flagged as an error by TypeScript

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

File: src/app/core/data/persistent-identifier-data.service.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * http://www.dspace.org/license/
+ */
+import { RequestService } from './request.service';
+import { RemoteDataBuildService } from '../cache/builders/remote-data-build.service';
+import { ObjectCacheService } from '../cache/object-cache.service';
+import { HALEndpointService } from '../shared/hal-endpoint.service';
+import { IdentifiableDataService } from './base/identifiable-data.service';
+import { DSpaceObject } from '../shared/dspace-object.model';
+
+export class PersistentIdentifierDataService extends IdentifiableDataService<DSpaceObject> {
```
Comment (tdonohue) [inline src/app/core/data/persistent-identifier-data.service.ts:15]: Can we add TypeDocs or a description to this new data service, please?

File: src/app/core/data/root-data.service.ts [hunk 2]
```diff
@@ -1,70 +1,33 @@
...
     protected objectCache: ObjectCacheService,
     protected halService: HALEndpointService,
-    protected notificationsService: NotificationsService,
-    protected http: HttpClient,
-    protected comparator: DefaultChangeAnalyzer<Root>,
-    protected restService: DspaceRestService) {
-    this.dataService = new DataServiceImpl(requestService, rdbService, null, objectCache, halService, notificationsService, http, comparator);
+    protected restService: DspaceRestService,
+  ) {
+    super('', requestService, rdbService, objectCache, halService, 6 * 60 * 60 * 1000);
```
Comment (tdonohue) [inline src/app/core/data/root-data.service.ts:30]: Could we either describe the `6 * 60 * 60 * 1000`, or change it back into a variable (it was previously named `responseMsToLive`)?  It would make this code easier to understand (and easier to replace with a config later if needed).

File: src/app/core/submission/vocabularies/vocabulary-entry-details.data.service.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import { RemoteData } from '../../data/remote-data';
+import { PaginatedList } from '../../data/paginated-list.model';
+import { SearchData, SearchDataImpl } from '../../data/base/search-data';
+import { Injectable } from '@angular/core';
+import { VOCABULARY_ENTRY_DETAIL } from './models/vocabularies.resource-type';
+import { dataService } from '../../data/base/data-service.decorator';
+
+@Injectable()
+@dataService(VOCABULARY_ENTRY_DETAIL)
+export class VocabularyEntryDetailsDataService extends IdentifiableDataService<VocabularyEntryDetail> implements FindAllData<VocabularyEntryDetail>, SearchData<VocabularyEntryDetail> {
```
Comment (tdonohue) [inline src/app/core/submission/vocabularies/vocabulary-entry-details.data.service.ts:27]: Please add TypeDocs/comments to describe this new data service

File: src/app/core/submission/vocabularies/vocabulary.data.service.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import { Observable } from 'rxjs';
+import { RemoteData } from '../../data/remote-data';
+import { PaginatedList } from '../../data/paginated-list.model';
+import { Injectable } from '@angular/core';
+import { VOCABULARY } from './models/vocabularies.resource-type';
+import { dataService } from '../../data/base/data-service.decorator';
+
+@Injectable()
+@dataService(VOCABULARY)
+export class VocabularyDataService extends IdentifiableDataService<Vocabulary> implements FindAllData<Vocabulary> {
```
Comment (tdonohue) [inline src/app/core/submission/vocabularies/vocabulary.data.service.ts:26]: Again, please add TypeDocs/comments to describe this new data service

**Review by tdonohue — COMMENTED** [review]
@ybnd : Thank you for this refactor!  I like how this really removes a lot of duplicative code from throughout existing data services. I also appreciate the _detailed_ TypeDocs in each of the new data service classes.

I've given this some thorough testing over the last few days & I'm not able to find any bugs so far. I've tried all major activities in the UI (submit, workflow, delete, edit, version, relationships, browse/search, etc) and everything seems to still be working great.

In reviewing the code, I see a lot of this PR is the same small refactors in many many classes. So it wasn't as difficult to review as I initially expected.

Overall the code looks good.  I would appreciate though adding new specs/tests to the newly created `src/app/core/data/base` data services where it's missing (more on that inline below) as these are now widely used, core services that need specs.  I also found a few places where TypeDocs were overlooked & a few other minor changes. 

In summary though, I'm nearly a +1.  All my feedback is minor in nature & doesn't impact the refactor itself.  Thanks again!

File: src/app/core/data/root-data.service.ts (hunk 2 above)
Comment (ybnd) [reply src/app/core/data/root-data.service.ts]: Agreed that putting the `responseMsToLive` in the constructor only is not as readable.
The same applies to `linkPath` and the new `constructIdEndpoint`.

I've changed the constructor signature a bit to move those back into the class body.
"Inner" services still override them in the constructor, since we need to set all of them to the same value for composed services to work properly.

Comment (ybnd) [reply src/app/core/data/root-data.service.ts]: As for the "6 hours":
* The root endpoint is not expected to change all that often; we're just caching it for an arbitrarily long while so that we don't spend time refreshing when we really don't need to
* If we make it shorter it should probably still be longer than the default 15 minutes

File: src/app/core/data/base/create-data.spec.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      halService,
+      notificationsService,
+    );
+  }
+
+  beforeEach(() => {
+    service = initTestService();
+  });
+
+  // todo: add specs (there were no ceate specs in original DataService suite!)
```
[the remark this answers is not in the document]
Comment (ybnd) [reply src/app/core/data/base/create-data.spec.ts:111]: @tdonohue I'd be happy to, but we would need to update the duration a bit.

* I think doing this thoroughly would take another **2-4h**, and we're already pushing 32h (not counting Art's time)
* Along with the CREATE/PUT/... specs I'd also like to add more specs for the actual composition, like [these ones ](https://github.com/DSpace/dspace-angular/pull/1791/files#diff-36a570953cdd4d885f59d700f963962fc631613bfecea4979b2a501af973cfaaR27)
  During the refactor I'd noticed that it's quite easy to overlook small mistakes when writing the "patch-through" methods; having these basic test methods we can reuse for each concrete service would give some peace of mind.
  This would be another **1-2h**.

**Comment by ybnd** [comment]
Last commit seems to have introduced some weird e2e failures, I'm looking into it
```
ERROR Error: Uncaught (in promise): TypeError: Cannot read properties of undefined (reading 'split')
TypeError: Cannot read properties of undefined (reading 'split')
    at HALEndpointService.getEndpoint (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:759070)
    at SearchDataImpl.getSearchEndpoint (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:323173)
    at SearchDataImpl.getSearchByHref (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:322943)
    ...
```

File: src/app/core/data/base/create-data.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tdonohue) [reply src/app/core/data/base/create-data.spec.ts:111]: @ybnd : Since it sounds like these additional tests will take less than a day (maybe only 1/2 a day), I'd like to go ahead with them in this PR, as I think it makes it easier to trust this refactor & provides more stability going forward. 

I'll bump the initial `30h` estimate to `38h`... which should at least give you 6 more hours to complete these tasks.  If the total still ends up over by a bit, that's OK. But, we obviously want to get as close to this estimate as possible. Thanks!

File: src/app/core/data/root-data.service.ts (hunk 2 above)
Comment (ybnd) [reply src/app/core/data/root-data.service.ts]: Regarding the constructor refactor ↑
* Looks like the e2e failures were due to the fact that I didn't account for TypeScript here, that refactor would never have worked (but still passed all specs!)
* I've reverted it, but have no other ideas as to how to make this more clear

File: src/app/core/data/base/search-data.spec.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      objectCache,
+      halService,
+    );
+  }
+
+  beforeEach(() => {
+    service = initTestService();
+  });
+
+  // todo: add specs (there were no search specs in original DataService suite!)
```
[the remark this answers is not in the document]
Comment (ybnd) [reply src/app/core/data/base/search-data.spec.ts]: Done.

File: src/app/core/data/bitstream-data.service.ts [hunk 7]
```diff
@@ -180,8 +183,89 @@ export class BitstreamDataService extends DataService<Bitstream> {
...
+   * Create the HREF for a specific object's search method with given options object
+   *
+   * @param searchMethod The search method for the object
+   * @param options The [[FindListOptions]] object
+   * @return {Observable<string>}
+   *    Return an observable that emits created HREF
+   * @param linksToFollow   List of {@link FollowLinkConfig} that indicate which {@link HALLink}s should be automatically resolved
+   */
+  public getSearchByHref(searchMethod: string, options?: FindListOptions, ...linksToFollow: FollowLinkConfig<Bitstream>[]): Observable<string> {
+    return this.searchData.getSearchByHref(searchMethod, options, ...linksToFollow);
```
[the remark this answers is not in the document]
Comment (ybnd) [reply src/app/core/data/bitstream-data.service.ts:201]: Tests for `SearchData` etc. are now in place.
I've also added reusable methods to test the "wiring" of delegated calls like these.

File: src/app/core/data/base/create-data.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ybnd) [reply src/app/core/data/base/create-data.spec.ts]: Added specs as discussed.

Simplified some of the `*Data` interface signatures to make them easier to test and implement
* No optional methods
* Promoted one of the previously optional method as it seemed useful
* Refactored out the usages of the other ones

**Comment by tdonohue** [comment]
@ybnd : You might already be aware of it, but this PR seems to be broken now...after the most recent refactors there are compilation errors.  See the automated CI tests for details.

**Review by atarix83 — APPROVED** [review]
thanks a lot @ybnd  for this PR

it's a huge refactoring, but all seems to be done very well. I've deeply tested the application and didn't find out any issue

**Review by tdonohue — APPROVED** [review]
👍 Thank you @ybnd !  I agree this is an impressive refactor.  Testing this today and not found any issues myself either.  I also appreciate all the additional tests added.

Merging this immediately, as it's at +2.... plus, I'm sure it will impact some other PRs
